File: server/python_server/prompt_shortcut.py
import re
import json

prompt_shortcut_dict = {}


def readToJson():
    return load()


def writeToJson(file_name, data_dict):
    with open(file_name, "w") as outfile:
        json.dump(data_dict, outfile, indent=4)


def load():
    global prompt_shortcut_dict
    try:
        with open("prompt_shortcut.json") as f_obj:
            data = json.load(f_obj)
            prompt_shortcut_dict = data
            print(data)
    except IOError:
        print("prompt_shortcut.json is not found")
    return prompt_shortcut_dict


def find_words_inside_braces(string):
    pattern = r"\{(.*?)\}"
    return re.findall(pattern, string)


# text = "a beautiful girl holding a cute cat {style_1} on sunny day"
# text = "a beautiful girl holding a cute cat {    style_1 } on sunny day"
text = "a beautiful girl{    }, {char1}, {painterly_style} holding a cute cat {    style_1 } on sunny day"
# text = "a beautiful girl {char1 } holding a cute cat on sunny day"


def replaceShortcut(text, prompt_shortcut_dict):
    raw_keywords = find_words_inside_braces(text)
    strip_keywords = [s.strip() for s in raw_keywords]
    original_substrings = ["{" + s + "}" for s in raw_keywords]

    print("raw_keywords: ", raw_keywords)
    print("strip_keywords: ", strip_keywords)
    print("original_substrings: ", original_substrings)

    for i, word in enumerate(strip_keywords):
        print("word: ", word)
        if word and word in prompt_shortcut_dict:
            prompt = prompt_shortcut_dict[word]
            print("prompt: ", prompt)
            text = text.replace(original_substrings[i], prompt)

    print("final text: ", text)
    return text

File: utility/sdapi/prompt_shortcut.js
function find_words_inside_braces(string) {
    const re = /\{(.*?)\}/g
    let keywords = string.match(re)
    // console.log('keywords: ', keywords)
    if (!keywords) {
        //avoid null keywords
        keywords = []
    }
    return keywords
}

function replaceShortcut(text, prompt_shortcut_json) {
    const original_keywords = find_words_inside_braces(text)
    const strip_keywords = original_keywords.map((s) => {
        let content = s.slice(1, -1) //remove '{' and  '}'
        content = content.trim() //remove any space in the beginning  and end of content
        return content
    })

    // original_substrings = list(map(lambda s: '{'+s+'}',raw_keywords))

    // print("strip_keywords: ", strip_keywords)
    // print("original_substrings: ",original_substrings)
    // # print ("text:",text)

    let i = 0
    for (const word of strip_keywords) {
        // # word = word.strip()
        //     print("word: ",word)
        if (word.length > 0 && prompt_shortcut_json.hasOwnProperty(word)) {
            const prompt = prompt_shortcut_json[word]
            console.log('prompt: ', prompt)
            text = text.replace(original_keywords[i], prompt)
        }
    }
    console.log('final text: ', text)
    return text
}
module.exports = {
    find_words_inside_braces,
    replaceShortcut,
}
